avocado/core/nrunner.py

```diff
@@ -613,97 +569,166 @@ def wait(self):
-    (("recipe", ),
-     {'type': str, 'help': 'Path to the task recipe file'}),
+        (("-a", "--arg"),
+         {'action': "append", 'default': [],
+          'help': 'Simple arguments to runnable'}),
+
+        (('kwargs',),
+         {'default': [], 'type': _parse_key_val, 'nargs': '*',
+          'metavar': 'KEY_VAL',
+          'help': 'Keyword (key=val) arguments to runnable'}),
     )
 
+    CMD_RUNNABLE_RUN_RECIPE_ARGS = (
+        (("recipe", ),
+         {'type': str, 'help': 'Path to the runnable recipe file'}),
+    )
 
-def subcommand_task_run_recipe(args, echo=print):
-    task = Task.from_recipe(args.get('recipe'))
-    task_run(task, echo)
+    CMD_TASK_RUN_ARGS = (
+        (("-i", "--identifier"),
+         {'type': str, 'required': True, 'help': 'Task unique identifier'}),
+        (("-s", "--status-uri"),
+         {'action': "append", 'default': None,
+          'help': 'URIs of status services to report to'}),
+    )
+    CMD_TASK_RUN_ARGS += CMD_RUNNABLE_RUN_ARGS
 
+    CMD_TASK_RUN_RECIPE_ARGS = (
+        (("recipe", ),
+         {'type': str, 'help': 'Path to the task recipe file'}),
+    )
 
-CMD_STATUS_SERVER_ARGS = (
-    (("uri", ),
-     {'type': str, 'help': 'URI to bind a status server to'}),
+    CMD_STATUS_SERVER_ARGS = (
+        (("uri", ),
+         {'type': str, 'help': 'URI to bind a status server to'}),
     )
 
+    def __init__(self, echo=print, prog=None, description=None):
+        self.echo = echo
+        self.parser = None
+        if prog is None:
+            prog = self.PROG_NAME
+        if description is None:
+            description = self.PROG_DESCRIPTION
+        self._setup_parser(prog, description)
+
+    def _setup_parser(self, prog, description):
+        self.parser = argparse.ArgumentParser(prog=prog,
+                                              description=description)
+        subcommands = self.parser.add_subparsers(dest='subcommand')
+        subcommands.required = True
+        for cmd_meth in self._get_commands_method_without_prefix():
+            attr = "CMD_%s_ARGS" % cmd_meth.upper()
+            cmd = cmd_meth.replace('_', '-')
+            cmd_parser = subcommands.add_parser(cmd)
+            if hasattr(self, attr):
+                for arg in getattr(self, attr):
+                    cmd_parser.add_argument(*arg[0], **arg[1])
+
+    def _get_commands_method_without_prefix(self):
+        prefix = 'command_'
+        return [c[0][len(prefix):]
+                for c in inspect.getmembers(self, inspect.ismethod)
+                if c[0].startswith(prefix)]
+
+    def run(self):
+        args = vars(self.parser.parse_args())
+        subcommand = args.get('subcommand')
+        if subcommand in self.get_commands():
+            meth_name = 'command_' + subcommand.replace('-', '_')
+            if hasattr(self, meth_name):
+                kallable = getattr(self, meth_name)
+                return kallable(args)
+
+    def get_commands(self):
```

Comment:
A public method, so, could you please provide a docstring, for here and other public methods?

Reply:
Yes, sure.